### PR TITLE
[2237] Course subjects and level tidy up

### PR DESCRIPTION
**Pull request by davidgisbey**
### Context
There's a large amount of code in our codebase related to UCAS subjects. This needs to either be removed or converted to use our new subjects.

### Changes proposed in this pull request
- Remove redundant code, such as the subject mapper service and ucas_subject model etc.
- Convert tests using UCAS subjects to use our new subjects. 
- Refactor copy course to new provider service. (this was causing validation errors due to previously not updating the start_date and applications_open from attribute within the new provider's cycle.

### Guidance fo review

Apologies for the length of the PR, but once the ucas_subject model was removed it caused a huge amount of test failures. It took all these changes to get them to pass.

I'll squash these commits down once changes are made the PR is approved.

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by davidgisbey** [comment]
[Course subjects related tidy up](https://trello.com/c/V22Wtcfy/2237-course-subjects-related-tidy-up)

File: app/models/dfe_subject.rb [hunk 1]
```diff
@@ -1,4 +1,3 @@
-  # TODO: chase up FINANCIAL_SUPPORT
```
Comment (defong) [inline app/models/dfe_subject.rb]: I think this one
- remove it entirely and double check that there is the 2020 values
- or just leave it

@dankmitchell @misaka was there a need to keep historical data?

File: spec/requests/api/v2/providers/courses/update_gcse_requirements_spec.rb [hunk 2]
```diff
@@ -141,15 +141,15 @@ def perform_request(gcse_requirements)
   end
 
   context "when not_set is provided on a secondary course" do
-    let(:secondary_subject) { build(:ucas_subject, :secondary) }
+    let(:secondary_subject) { build(:subject, type: 'SecondarySubject') }
```
Comment (defong) [inline spec/requests/api/v2/providers/courses/update_gcse_requirements_spec.rb]: can we have that as a trait or use one of the existing please

File: app/models/dfe_subject.rb (hunk 1 above)
Comment (misaka) [reply app/models/dfe_subject.rb]: There was some talk about potentially exposing historical course info? Guess we just need to make a call on that.

**Comment by misaka** [comment]
😻 

```
 65 files changed, 348 insertions(+), 1262 deletions(-)
```

File: spec/services/courses/copy_to_provider_service_spec.rb [hunk 3]
```diff
@@ -33,14 +34,23 @@
...
     expect(new_course.content_status).to eq :rolled_over
     expect(new_course.ucas_status).to eq :new
     expect(new_course.open_for_applications?).to be_falsey
   end
 
+  it "updates the applications_open_from and start date attributes" do
+    service.execute(course: course, new_provider: new_provider)
+    expect(new_course.start_date).to eq course.start_date + 1.year
+    expect(new_course.applications_open_from).to eq course.applications_open_from + 1.year
+  end
```
Comment (misaka) [inline spec/services/courses/copy_to_provider_service_spec.rb]: Is this related to subjects? Or is it just along for the ride?

File: app/services/courses/copy_to_provider_service.rb [hunk 4]
```diff
@@ -13,10 +13,12 @@ def execute(course:, new_provider:)
       Course.transaction do
         new_course = course.dup
         new_course.provider = new_provider
+        year_differncial = new_course.recruitment_cycle.year.to_i - course.recruitment_cycle.year.to_i
```
Comment (misaka) [inline app/services/courses/copy_to_provider_service.rb]: "year_differncial"? Also, why in this PR?

File: app/models/dfe_subject.rb (hunk 1 above)
Comment (dankmitchell) [reply app/models/dfe_subject.rb]: TLDR; SHIP IT 🚢

I purpose we move this data into the new "Finacial incentives" database table as I think we want to retain historical data however that's out of scope for this PR and as it's in source control we can pull it back whenever we need to 👌

File: spec/factories/subjects.rb [hunk 5]
```diff
@@ -16,7 +16,6 @@
     type {
       %w[PrimarySubject
          SecondarySubject
-         ModernLanguagesSubject
```
Comment (defong) [inline spec/factories/subjects.rb]: why was that removed?

File: spec/factories/subjects.rb [hunk 6]
```diff
@@ -16,7 +16,6 @@
     type {
```
Comment (defong) [inline spec/factories/subjects.rb]: can we do any thing nice with

`SubjectCreatorService.new.execute` in the factory

given that the its no longer unknown...

File: spec/models/course_spec.rb [hunk 7]
```diff
@@ -695,38 +695,13 @@
     end
   end
 
-  context "UCAS level (deprecated)" do
-    context "with no ucas_subjects" do
-      subject { create(:course) }
-      its(:ucas_level) { should eq(:secondary) }
-    end
-
-    context "with primary ucas_subjects" do
```
Comment (defong) [inline spec/models/course_spec.rb]: deleted too much?

given a `course`
when it has that `level`
then `gcse_subjects_required` is set correctly?

File: spec/models/course_spec.rb [hunk 8]
```diff
@@ -695,38 +695,13 @@
...
-      subject { create(:course, level: "primary", ucas_subjects: [find_or_create(:ucas_subject, :primary)]) }
-      its(:ucas_level) { should eq(:primary) }
-      its(:gcse_subjects_required) { should eq(%w[maths english science]) }
-    end
-
-    context "with secondary ucas_subjects" do
```
Comment (defong) [inline spec/models/course_spec.rb]: deleted too much?

given a `course`
when it has that `level`
then `gcse_subjects_required` is set correctly?

File: spec/models/course_spec.rb [hunk 9]
```diff
@@ -695,38 +695,13 @@
...
-      subject { create(:course, level: "secondary", ucas_subjects: [find_or_create(:ucas_subject, subject_name: "physical education")]) }
-      its(:ucas_level) { should eq(:secondary) }
-      its(:gcse_subjects_required) { should eq(%w[maths english]) }
-    end
-
-    context "with further education ucas_subjects" do
```
Comment (defong) [inline spec/models/course_spec.rb]: deleted too much?

given a `course`
when it has that `level`
then `gcse_subjects_required` is set correctly?

File: spec/README.md [hunk 10]
```diff
@@ -36,8 +36,8 @@ let(:provider) { course.provider }
 ## Overriding the creation of a dependent object
 
 ```
-let(:english_subject) { create :ucas_subject, :english }
-let(:course)          { create :course, ucas_subjects: [english_subject] }
+let(:english_subject) { create :subject, :english }
+let(:course)          { create :course, subjects: [english_subject] }
```
Comment (defong) [inline spec/README.md]: probably should `level` it

File: spec/serializers/api/v2/serializable_course_spec.rb [hunk 11]
```diff
@@ -156,7 +156,7 @@
     let(:course) { create(:course, subjects: subjects) }
 
     describe "are taken from the course" do
-      let(:subjects) { [find_or_create(:ucas_subject, :primary)] }
+      let(:subjects) { [find_or_create(:subject, :primary)] }
```
Comment (defong) [inline spec/serializers/api/v2/serializable_course_spec.rb]: :eyes: at the top

```ruby
  # TODO: level now drives the valid subjects that can be assigned to a
  #       given course
  # TODO: bursary and scholarship info should now live in the database
  # TODO: chase up FINANCIAL_SUPPORT
xcontext "subjects & level" do
```

File: spec/factories/subjects.rb [hunk 12]
```diff
@@ -16,7 +16,6 @@
...
       %w[PrimarySubject
```
Comment (defong) [inline spec/factories/subjects.rb]: :eyes: at the top

```ruby
  # TODO: use known subject list
  factory :subject do
```

File: spec/README.md (hunk 10 above)
Comment (misaka) [reply spec/README.md]: I think @defong means add a trait like `primary`.

**Comment by misaka** [comment]
> 1.
>
>  _ruby code proving we do a lookup of the modern languages subject by name_
>
> do we need to consider indexing subject_name and type?

I dunno. I think that with the cardinality here, having an index vs forcing a table scan is not going to be appreciably faster, if at all.

**Comment by misaka** [comment]
> 2.
>
> ```ruby
> let(:course) { create(:course, level: "primary", subjects: [subjects]) }
>  let(:course) { create(:course, level: "further_education", subjects: [subjects]) }
> let(:course) { create(:course, level: "secondary", subjects: [subjects]) }
> ```
> 
> can we do anything nice, by vitue of creating a `course` and its `level` is already inferred?

Do you mean if we give it specific subjects, can we infer the level in the `course` factory? Yes, we can have an `after_build` that examines the subjects the course has, and changes the level accordingly. It does get a little hairy, though, because when you do that you might wipe out the `level` that the user is setting ... and how do we know if the user specified `level` or `subjects`?

**Comment by misaka** [comment]
> 3.
>
> migration to drop ucas_subjects?

🚀

**Comment by defong** [comment]
> Do you mean if we give it specific subjects, can we infer the level in the `course` factory? Yes, we can have an `after_build` that examines the subjects the course has, and changes the level accordingly. It does get a little hairy, though, because when you do that you might wipe out the `level` that the user is setting ... and how do we know if the user specified `level` or `subjects`?

was thinking along

`let(:course) { create(:course, :primary" }`
`let(:course) { create(:course, :secondary }`
`let(:course) { create(:course, :secondary_with_multiple_subjects }`

etc

File: spec/services/courses/copy_to_provider_service_spec.rb (hunk 3 above)
Comment (davidgisbey) [reply spec/services/courses/copy_to_provider_service_spec.rb]: As discussed it was not updating the new provider's cycle and has snuck in. Will fix the typo.

File: spec/factories/subjects.rb (hunk 5 above)
Comment (davidgisbey) [reply spec/factories/subjects.rb]: Exactly, wasn't sure if we were doing after builds etc. anymore. It triggers a validation as it does not create a modern language subject which is required.

File: spec/requests/api/v2/providers/courses/update_gcse_requirements_spec.rb (hunk 2 above)
Comment (davidgisbey) [reply spec/requests/api/v2/providers/courses/update_gcse_requirements_spec.rb]: Done 👍

File: spec/models/course_spec.rb (hunk 7 above)
Comment (davidgisbey) [reply spec/models/course_spec.rb]: discussed. is tested elsewhere

File: spec/README.md (hunk 10 above)
Comment (davidgisbey) [reply spec/README.md]: Going to do an additional PR to do this in the factories.

File: spec/factories/subjects.rb (hunk 12 above)
Comment (davidgisbey) [reply spec/factories/subjects.rb]: Will do in a follow up pr

File: app/models/dfe_subject.rb (hunk 1 above)
Comment (davidgisbey) [reply app/models/dfe_subject.rb]: Going to do this in a follow up PR 👍

File: spec/factories/subjects.rb (hunk 5 above)
Comment (davidgisbey) [reply spec/factories/subjects.rb]: Have noted this to look at in a follow up card.

File: spec/serializers/api/v2/serializable_course_spec.rb (hunk 11 above)
Comment (davidgisbey) [reply spec/serializers/api/v2/serializable_course_spec.rb]: Have noted for a future pr
